### PR TITLE
Added support for some of the missing joint parameters

### PR DESCRIPTION
**Pull request by mihe**
Related to jrouwe/JoltPhysics#555.

This PR aims to add support for some of the missing joint parameters, like springs and soft limits.

Here's a complete list of the parameters/flags that are currently missing, as well as what their current status is:

## Legend

- ✅ Supported in this PR
- ⏳️ Not yet supported in this PR
- 🚫 Won't be supported in this PR
- ❓ Unknown/uncertain

## PinJoint3D

- 🚫 `PIN_JOINT_BIAS`
- 🚫 `PIN_JOINT_DAMPING`
- 🚫 `PIN_JOINT_IMPULSE_CLAMP`

## HingeJoint3D

- 🚫 `HINGE_JOINT_BIAS`
- ✅ `HINGE_JOINT_LIMIT_BIAS`
- 🚫 `HINGE_JOINT_LIMIT_SOFTNESS`
- 🚫 `HINGE_JOINT_LIMIT_RELAXATION`

## SliderJoint3D

- ⏳️ `SLIDER_JOINT_LINEAR_LIMIT_SOFTNESS`
- ⏳️ `SLIDER_JOINT_LINEAR_LIMIT_RESTITUTION`
- ⏳️ `SLIDER_JOINT_LINEAR_LIMIT_DAMPING`
- ⏳️ `SLIDER_JOINT_LINEAR_MOTION_SOFTNESS`
- ⏳️ `SLIDER_JOINT_LINEAR_MOTION_RESTITUTION`
- ⏳️ `SLIDER_JOINT_LINEAR_MOTION_DAMPING`
- 🚫 `SLIDER_JOINT_LINEAR_ORTHOGONAL_SOFTNESS`
- 🚫 `SLIDER_JOINT_LINEAR_ORTHOGONAL_RESTITUTION`
- 🚫 `SLIDER_JOINT_LINEAR_ORTHOGONAL_DAMPING`
- 🚫 `SLIDER_JOINT_ANGULAR_LIMIT_UPPER`
- 🚫 `SLIDER_JOINT_ANGULAR_LIMIT_LOWER`
- 🚫 `SLIDER_JOINT_ANGULAR_LIMIT_SOFTNESS`
- 🚫 `SLIDER_JOINT_ANGULAR_LIMIT_RESTITUTION`
- 🚫 `SLIDER_JOINT_ANGULAR_LIMIT_DAMPING`
- 🚫 `SLIDER_JOINT_ANGULAR_MOTION_SOFTNESS`
- 🚫 `SLIDER_JOINT_ANGULAR_MOTION_RESTITUTION`
- 🚫 `SLIDER_JOINT_ANGULAR_MOTION_DAMPING`
- 🚫 `SLIDER_JOINT_ANGULAR_ORTHOGONAL_SOFTNESS`
- 🚫 `SLIDER_JOINT_ANGULAR_ORTHOGONAL_RESTITUTION`
- 🚫 `SLIDER_JOINT_ANGULAR_ORTHOGONAL_DAMPING`

## ConeTwistJoint3D

- 🚫 `CONE_TWIST_JOINT_BIAS`
- 🚫 `CONE_TWIST_JOINT_SOFTNESS`
- 🚫 `CONE_TWIST_JOINT_RELAXATION`

## Generic6DOFJoint

- ⏳️ `G6DOF_JOINT_LINEAR_LIMIT_SOFTNESS`
- ⏳️ `G6DOF_JOINT_LINEAR_RESTITUTION`
- ⏳️ `G6DOF_JOINT_LINEAR_DAMPING`
- 🚫 `G6DOF_JOINT_ANGULAR_LIMIT_SOFTNESS`
- 🚫 `G6DOF_JOINT_ANGULAR_DAMPING`
- 🚫 `G6DOF_JOINT_ANGULAR_RESTITUTION`
- 🚫 `G6DOF_JOINT_ANGULAR_FORCE_LIMIT`
- 🚫 `G6DOF_JOINT_ANGULAR_ERP`

CC: @jrouwe

### CODE REVIEW

**Comment by mihe** [comment]
I'm not sure if you'd prefer to keep the conversation in the issue on your end, @jrouwe, or if it's alright to do it here as well, but when going through the list of parameters I was unsure what the deal was with `HINGE_JOINT_LIMIT_BIAS` and `G6DOF_JOINT_ANGULAR_FORCE_LIMIT`.

You mentioned that the hinge should probably be implemented using the "softness * relaxation and softness * damping" method, but there's no damping there as far as I can tell, unless `HINGE_JOINT_LIMIT_BIAS` is the damping somehow.

I don't think we touched on `G6DOF_JOINT_ANGULAR_FORCE_LIMIT`. It's described as "The maximum amount of force that can occur, when rotating around the axes." in Godot's documentation.

(Also, forgive this one-off force push, I'll try to keep the commit history somewhat sane to allow for easier incremental reviews.)

**Comment by jrouwe** [comment]
Hello,

As far as I can see HINGE_JOINT_LIMIT_BIAS = stiffness and HINGE_JOINT_LIMIT_RELAXATION = damping (in the current godot version that ignores softness, if you want to do it correctly I think you need to multiply both with the softness) 

I will look into G6DOF_JOINT_ANGULAR_FORCE_LIMIT and your changes tomorrow.

**Comment by mihe** [comment]
I pulled out some of the refactoring that had no business being in this PR to begin with, which necessitated another force-push. Hopefully that should make the changes a lot easier to grok.

I'll be adding the changes to `HingeJoint3D` later today.

**Review by jrouwe — APPROVED** [review]
Looks good to me!

File: src/joints/jolt_generic_6dof_joint_impl_3d.cpp [hunk 1]
```diff
@@ -530,12 +501,27 @@ void JoltGeneric6DOFJointImpl3D::update_motor_limit(int32_t p_axis) {
 	}
 }
 
+void JoltGeneric6DOFJointImpl3D::update_spring_stiffness(int32_t p_axis) {
+	if (auto* constraint = static_cast<JPH::SixDOFConstraint*>(jolt_ref.GetPtr())) {
+		JPH::MotorSettings& motor_settings = constraint->GetMotorSettings((JoltAxis)p_axis);
+		motor_settings.mSpringSettings.mFrequency = (float)spring_stiffness[p_axis];
```
[the remark this answers is not in the document]
Comment (mihe) [reply src/joints/jolt_generic_6dof_joint_impl_3d.cpp]: Ah, I didn't see the union. Thank you!

**Comment by jrouwe** [comment]
Regarding G6DOF_JOINT_ANGULAR_FORCE_LIMIT and G6DOF_JOINT_ANGULAR_MOTOR_FORCE_LIMIT:

What the code does is to first detect if the motor is [outside of the angular imits](https://github.com/godotengine/godot/blob/72b59325cf7beba7e6e9170cf6023a079fd58672/servers/physics_3d/joints/godot_generic_6dof_joint_3d.cpp#L64C13-L80) and use that to calculate the actual [angular impulse](https://github.com/godotengine/godot/blob/72b59325cf7beba7e6e9170cf6023a079fd58672/servers/physics_3d/joints/godot_generic_6dof_joint_3d.cpp#L90-L97)

So basically:

* If G6DOF_JOINT_ANGULAR_MOTOR_TARGET_VELOCITY is set then G6DOF_JOINT_ANGULAR_MOTOR_FORCE_LIMIT is used as the motor force limit. 
* If, however, the limits are violated, the motor is turned off and instead a spring to enforce the limits is activated. G6DOF_JOINT_ANGULAR_FORCE_LIMIT is the max force that it will apply in that case.

Jolt doesn't really work that way as it will keep the motor active and enforce the limits at the same time (meaning the motor will push until the limit force is big enough to counteract the motor). I personally think that that is more consistent as turning the motor off will give a discontinuity: the motor will push the constraint past the limits, instantly turn off, spring returns the constraint within the limits again, motor instantly turns on, pushes constraints past limits etc.

What goes where:
* G6DOF_JOINT_ANGULAR_MOTOR_FORCE_LIMIT - this goes into the motor settings for the velocity motor. You've already implemented this in the right way.
* G6DOF_JOINT_ANGULAR_FORCE_LIMIT - Jolt doesn't support a min/max force for limit springs yet. It's not hard but I feel it's quite a fringe case where you would want to use this.

**Comment by mihe** [comment]
Great, thank you for taking a closer look at it. I agree that the inconsistency there seems a bit jarring. It sounds like `G6DOF_JOINT_ANGULAR_FORCE_LIMIT` is staying unsupported then.

**Comment by mihe** [comment]
I've been treading water with this for a bit now, as the soft limits for `HingeJoint3D` are proving difficult for me to translate.

For reference, here are the ranges (enforced in the editor) and default values (which I sadly can't change) of these properties:

- `HINGE_JOINT_LIMIT_BIAS`
    - Min: 0.01
    - Max: 0.99
    - Step: 0.01
    - Default: 0.3
- `HINGE_JOINT_LIMIT_RELAXATION`
    - Min: 0.01
    - Max: 16.0
    - Step: 0.01
    - Default: 1.0
- `HINGE_JOINT_LIMIT_SOFTNESS`
    - Min: 0.01
    - Max: 16.0
    - Step: 0.01
    - Default: 0.9

So `HINGE_JOINT_LIMIT_BIAS` is a factor in both Godot 3 and Godot 4, and I'm not entirely sure how I can map this to Jolt's stiffness. Multiplying it by some arbitrary value, which after some experimentation seems to be in the millions, and then passing it on as `mStiffness`, makes it behave close-ish to what I can see in Bullet, as this factor doesn't seem immune to changes in mass there either. Once you start messing with the damping this "calibration" sort of goes out the window though.

`HINGE_JOINT_LIMIT_RELAXATION` is apparently described in Godot's documentation as "The lower this value, the more rotation gets slowed down", which to me sounds like the inverse of how damping behaves, and seems to reflect how it behaves at runtime in Bullet. This confuses me quite a bit, considering the name and value range of this property, so I'm again not sure how to map this to Jolt's damping.

Could I be missing something here, or is this just an unfortunate discrepancy between these two worlds?

It wouldn't be the end of the world to leave these as unsupported and instead implement them for the eventual custom nodes later instead.

**Comment by phelioz** [comment]
Unfortunately, I don't have much to contribute to this question. However, I wanted to express that, from my perspective, I would definitely utilize the "custom Jolt nodes" if they were available. It would be wonderful to be able to use frequency and damping directly and to have nodes that map more directly to the Jolt joint constraints.

**Comment by jrouwe** [comment]
I see I didn't look closely enough:

```
real_t amplitude = ((angVelB - angVelA).dot(axisA) * m_relaxationFactor + m_correction * (real_t(1.) / p_step) * m_biasFactor) * m_limitSign;

real_t impulseMag = amplitude * m_kHinge;
```

or 

```
torque_impulse = -m_kHinge * m_biasFactor / p_step * x - m_kHinge * m_relaxationFactor * v
```

so

```
torque = -m_kHinge * m_biasFactor / p_step^2 * x - m_kHinge * m_relaxationFactor / p_step * v
```

so

```
jolt_stiffness = m_kHinge * m_biasFactor / p_step^2
jolt_damping = m_kHinge * m_relaxationFactor / p_step
```

m_kHinge is calculated like this during setup:

```
Vector3 axisA = A->get_transform().basis.xform(m_rbAFrame.basis.get_column(2));
m_kHinge = 1.0f / (A->compute_angular_impulse_denominator(axisA) + B->compute_angular_impulse_denominator(axisA));
```

So a function of the hinge axis and the inertias of both bodies.

These formulas are actually very close to what Jolt uses for the frequency and damping and I think:

```
(2 * pi * frequency)^2 = m_biasFactor / p_step^2
4 * pi * frequency * damping = m_relaxationFactor / p_step
```

So I think you can also keep the spring in freq/damping mode and calculate the frequency and damping using these formulas.

(and I probably made a math mistake somewhere here)

**Comment by jrouwe** [comment]
Hello,

I've been going over the math again and I cannot spot my mistake. If the godot version doesn't behave well then I can understand that Jolt also doesn't if it uses the same math.

I looked at the Bullet source code and that implementation seems to be completely different (I find it hard to follow though). It looks like the formula goes like:

```
torque_impulse = -m_kHinge * m_biasFactor * max(0.2 / p_step * x,  m_relaxationFactor * v) - m_kHinge * 1.0 * v
```

(I've simplified this a bit by omitting some details when hitting the lower vs the upper limit)

The relaxation factor is only used when the velocity v is bigger than the position error x (multiplied by some constants). The purpose of the relaxation factor seems to be to make the hinge bounce against the limits (equivalent of the restitution). The damping factor is hardcoded to 1 (m_damping is not set in the block of code you mentioned and when debugging I saw it defaults to 1) and the 0.2 is the global ERP factor which is also not overridden by godot.

I think we have to leave out relaxation as Jolt doesn't support that max operation, that leaves us with:

```
jolt_stiffness = m_kHinge * m_biasFactor * 0.2 / p_step^2
jolt_damping = m_kHinge / p_step
```

And in frequency damping terms:

```
(2 * pi * frequency)^2 = m_biasFactor * 0.2 / p_step^2
4 * pi * frequency * damping = 1 / p_step
```

or:

```
frequency = sqrt(m_biasFactor) / (2 * pi * sqrt(5) * p_step)
damping = 1 / (4 * pi * frequency * p_step)
```

I have no idea how accurate this is going to be....

**Comment by jrouwe** [comment]
Thinking about it some more, I think we could also approximate max(a, b) with a + b to get:

```
jolt_damping = m_kHinge * (1 + m_relaxationFactor) / p_step
```

or:

```
damping = (1 + m_relaxationFactor) / (4 * pi * frequency * p_step)
```

**Comment by mihe** [comment]
## Formula 2

```
frequency = sqrt(m_biasFactor) / (2 * pi * sqrt(5) * p_step)
damping = 1 / (4 * pi * frequency * p_step)
```

This one is actually kinda scary how close it is to Bullet. I had to double-check to make sure I hadn't mixed up the files at first. So I would say you nailed the accuracy in terms of matching Bullet at least.

There looks to be a bit more bounce at the limits in the Bullet one, which I guess might be explained by the missing relaxation component in this particular formula, but honestly this would be Good Enough™ in my book. I had no illusions about perfect parity, and this is better than I could have imagined.

### Bias

https://github.com/godot-jolt/godot-jolt/assets/4884246/d84beeef-1484-4f3f-90ee-5e67e76590f6

### Relaxation

(No video showing relaxation since that isn't incorporated.)

## Formula 3

```
frequency = sqrt(m_biasFactor) / (2 * pi * sqrt(5) * p_step)
damping = (1 + m_relaxationFactor) / (4 * pi * frequency * p_step)
```

The relaxation here seemingly ends up inverted again, where a higher relaxation erroneously results in more damping, which ends up affecting both tests by the looks of it.

### Bias

https://github.com/godot-jolt/godot-jolt/assets/4884246/e173728f-3691-4160-882b-0ce45c45430f

### Relaxation

https://github.com/godot-jolt/godot-jolt/assets/4884246/b138a2aa-019b-43b8-a46a-38c4101bad6d

## Formula 4

```
frequency = sqrt(m_biasFactor) / (2 * pi * sqrt(5) * p_step)
magic = 1 - (m_relaxationFactor / 16)
damping = magic / (4 * pi * frequency * p_step)
```

This is a caveman attempt on my part, with no real basis in reality, but I was curious to see what it would look like if the dividend somehow ended up being an inverse factor of the relaxation.

I think the result are somewhat interesting. The bias test stays close to the results of formula 2, while still ending up with a relaxation that achieves something resembling an inverse damping, even if it doesn't really look anything like what Bullet produces.

I have no idea how you'd achieve this "correctly" though.

### Bias

https://github.com/godot-jolt/godot-jolt/assets/4884246/68439c48-2f66-4549-8a3f-53ea5dbec89b

### Relaxation

https://github.com/godot-jolt/godot-jolt/assets/4884246/ccdee8aa-84ca-4402-9cdb-73a32b24b777

## Summary

All-in-all, I'd be perfectly fine with settling for formula 2 and just leaving relaxation unsupported, unless you have more ideas.

**Comment by jrouwe** [comment]
I think you should go for formula 2 indeed. I think the fact that the relaxation isn't working is because I oversimplified the formula: The relaxation is only applied when the body is moving towards the limit and not away, this is not something you can replicate with the current Jolt implementation.

I cannot find any documentation (neither in godot, bullet or anywhere else on the internet) what the physical meaning of 'bias' and 'relaxation' is, so to me these are just confusing numbers. 

I also looked at PhysX's PxRevoluteJoint [limit settings](https://nvidia-omniverse.github.io/PhysX/physx/5.1.3/_build/physx/latest/class_px_joint_angular_limit_pair.html#_CPPv423PxJointAngularLimitPair) and they use stiffness and damping just like Jolt but have an extra 'restitution' property to make limits bouncy. Maybe I'll add that too one day (and maybe it can be related to relaxation somehow).

**Comment by mihe** [comment]
Apologies for letting this PR sit inactive for so long. With summer rolling around I've had to slow things down a bit, which has made more time-consuming tasks like this one tricky to fit into my short bursts of work.

With that said, I've also felt very conflicted about how to proceed with these changes.

To start with, it turned out that formula 2 for the hinge joint soft limits didn't actually work as well as I thought it did. By some stroke of luck it turned out that Jolt (with formula 2) and Bullet just happened to align in behavior at the particular timestep (240 hz) that I was using in the test project. Once I changed that to something like 60 hz they deviated from eachother quite a bit. This was however seemingly fixable by simply hardcoding the timestep into the formula, so that it became something more like to this:

```
frequency = sqrt(m_biasFactor) / (2 * pi * sqrt(5) * (1 / 240))
damping = 1 / (4 * pi * frequency * (1 / 240))
```

After some further tweaking to this (now magic) constant, to better match Bullet, I ended up with something closer to `(1 / 300)`. Obviously this change didn't feel very sensible though.

On top of this I've spent a little bit of time in the last few weeks trying to make sense of the (linear) soft limits for the slider and 6DOF joints, using a similar test project as what I used for the hinge limits, and I can safely say at this point that I don't understand them even a little bit. Out of the three parameters (softness, restitution and damping) the only one that doesn't seem to explode the joint at anything but the default value is softness. This applies to both Bullet and Godot Physics, weirdly enough.

With all that in mind, as well the whole thing about these parameters not seeming to have any real physical meaning, I'm feeling much less inclined to actually add support for the default joints' soft limits. I'm also worried that they might do more harm than good, since there's no way of actually disabling the soft limits with the parameter ranges that Godot uses. So all hinges, sliders and 6DOF joints would effectively be using soft limits at all times, which I feel demands that they are sensible and reliable. Those are not words I would use to describe their current state in either of the two reference implementations.

Once Godot 4.1 is released in the coming days that will (hopefully) unlock some things that I've been waiting for in order to implement the custom Jolt joints properly, so I will likely be doing a focused effort in the coming few weeks to add all that stuff, as well as figuring out some solution for the C# part of it. These will obviously use the more sensible frequency/damping interface for springs and soft limits, as well as add other useful features, like breakable joints and the ability to override the solver iterations on a per-joint basis.

I've gone ahead and pulled out the 6DOF springs from this PR (#478) and instead replaced it with the code for the hinge joint stuff, in case anyone is curious what that looked like. I also added support for spring equilibrium (target position/rotation) as part of #478, which seemed to work well enough. I'm not entirely sure if that rotation order thing is anything to worry about, but seeing as how Godot Physics doesn't implement this at all right now there's technically no compatibility to worry about anyway.

Anyway, please let me know if you have any objections to abandoning the soft limits for the default Godot joints, @jrouwe. Clearly this aspect of the Jolt integration has proven to be beyond my capabilities anyway, so I'm at the mercy of whatever support you can offer either way.

**Comment by jrouwe** [comment]
Thanks for the update! I will take another look at it. I know the creator of Bullet, so I may just send him an e-mail to ask what these numbers are supposed to mean.

**Comment by jrouwe** [comment]
I've been looking at this again, and I think you're right that we should just leave these parameters for now and have Jolt use hard limits for the default constraints. The parameters seem to be inconsistently implemented between Godot 3, 4 and Bullet and across the different constraint types, so it is going to be very difficult to implement this.

**Comment by mihe** [comment]
Thank you for taking yet another look at it. I really appreciate it.

It seems like 4.1 is [right around the corner](https://github.com/godotengine/godot/commit/970459615f6b2b4151742ec6d7ef8559f87fd5c5), so I might take a crack at the custom nodes this week even. Once I get going I would imagine they'll be implemented in fairly quick succession.

I'll go ahead and close this PR. Thanks again!
